Framework/Algorithms/src/CalcCountRate.cpp

```diff
@@ -0,0 +1,653 @@
+  m_visWs->setTitle(visWSName);
+
+  double Xmax = m_XRangeMax;
+  // a bit dodgy code. It can be generalized
+  if (std::isinf(Xmax)) {
+    auto Xbin = m_workingWS->readX(0);
+    for (int64_t i = Xbin.size() - 1; i >= 0; --i) {
+      if (!std::isinf(Xbin[i])) {
+        Xmax = Xbin[i];
+        break;
+      }
+    }
+    if (std::isinf(Xmax)) {
+      g_log.warning() << "All X-range for visualization workspace is infinity. "
+                         "Can not build visualization workspace in the units "
+                         "requested\n";
+      m_visWs.reset();
+      m_doVis = false;
+      return;
+    }
+  }
+
+  // define X-axis in target units
+  double dX = (Xmax - m_XRangeMin) / numXBins;
+  std::vector<double> xx(numXBins);
+  for (int i = 0; i < numXBins; ++i) {
+    xx[i] = m_XRangeMin + (0.5 + static_cast<double>(i)) * dX;
+  }
+  auto ax0 = new API::NumericAxis(xx);
+  ax0->setUnit(RangeUnits);
+  m_visWs->replaceAxis(0, ax0);
+
+  // define Y axis (in seconds);
+  double dt = (static_cast<double>(m_TRangeMax.totalNanoseconds() -
+                                   m_TRangeMin.totalNanoseconds()) /
+               static_cast<double>(numTBins)) *
+              1.e-9;
+  xx.resize(numTBins);
+  for (int i = 0; i < numTBins; i++) {
+    xx[i] = (0.5 + static_cast<double>(i)) * dt;
+  }
+  auto ax1 = new API::NumericAxis(xx);
+  auto labelY = boost::dynamic_pointer_cast<Kernel::Units::Label>(
+      Kernel::UnitFactory::Instance().create("Label"));
+  labelY->setLabel("sec");
+  ax1->unit() = labelY;
+  m_visWs->replaceAxis(1, ax1);
+
+  setProperty("VisualizationWs", m_visWs);
+
+  // define binning parameters used while calculating visualization
+  m_visX0 = m_XRangeMin;
+  m_visDX = dX;
+  m_visT0 = static_cast<double>(m_TRangeMin.totalNanoseconds());
+  m_visTmax = static_cast<double>(m_TRangeMax.totalNanoseconds());
+  m_visDT = (m_visTmax - m_visT0) / static_cast<double>(numTBins);
+
+  if (this->notmalizeCountRate()) {
+    m_visNorm.resize(numTBins);
+    this->buildVisWSNormalization(m_visNorm);
+  }
+}
+/** Helper function to check if visualization workspace should be used*/
+bool CalcCountRate::buildVisWS() const { return m_doVis; }
+
+/** Helper function, mainly for testing
+* @return  true if count rate should be normalized and false
+* otherwise */
+bool CalcCountRate::notmalizeCountRate() const { return m_normalizeResult; }
+/** Helper function, mainly for testing
+* @return  true if log derivative is used instead of log itself */
+bool CalcCountRate::useLogDerivative() const { return m_useLogDerivative; }
+
+/** method to prepare normalization vector for the visuzliation workspace using
+* data from normalization log with, usually, different number of time steps
+* Here we assume that the numner of time points in the visualization workspace
+* is smaller or equal to the number of points in the normalization log.
+*/
+void CalcCountRate::buildVisWSNormalization(
+    std::vector<double> &normalization) {
```

Comment:
Why do you need to pass the `normalization` parameter? It seems to be unused here.